packages/loopring_v3/contracts/impl/BasicDepositContract.sol

```diff
@@ -110,12 +113,17 @@ contract BasicDepositContract is IDepositContract, ReentrancyGuard
         override
         payable
         onlyExchange
+        nonReentrant
+        ifNonZero(amount)
     {
-        // Keep track how many tokens are deposited in the exchange
-        exchangeBalance[token] = exchangeBalance[token].sub(uint(amount));
-
-        // Transfer the tokens from the contract to the recipient
-        transferOut(to, token, amount);
+        if (isETHInternal(token)) {
+            to.sendETHAndVerify(amount, gasleft());
+        } else {
+            // Just in case the balance is unexpectedly smaller
+            uint balance = ERC20(token).balanceOf(address(this));
+            uint _amount = balance < amount ? balance : amount;
+            token.safeTransferAndVerify(to, _amount);
+        }
```

Comment:
I think we can make this a bit more efficient by using a try/catch. Now we always check the balance which will increase gas costs ~2,500 gas or something like that for each withdrawal. We can just always do the transfer, and only when it fails do this additional logic to cap it to the max available.

Another option is to let it fail, and let the user withdraw with `withdrawFromApprovedWithdrawals`, but we modify the function to take an extra amountToWithdraw parameter so it can be set to whatever is still available in the deposit contract. This way we never have to worry about this case onchain, at least not in this simple use of deposit contracts.

If we do this check for ERC20 tokens, why not for ETH?

Reply:
```solidity
try token.safeTransferAndVerify(to, amount) {} catch {
                uint balance = ERC20(token).balanceOf(address(this));
                require(balance < amount, "UNEXPECTED_STATE");
                token.safeTransferAndVerify(to,balance);
            }
```
we get:
```
Try can only be used with external function calls and contract creation calls.
```

Reply:
> If we do this check for ERC20 tokens, why not for ETH?
Because this contract can never send ETH out except by calling the withdrawal function.

Reply:
> Another option is to let it fail, and let the user withdraw with withdrawFromApprovedWithdrawals, but we modify the function to take an extra amountToWithdraw parameter so it can be set to whatever is still available in the deposit contract. This way we never have to worry about this case onchain, at least not in this simple use of deposit contracts.

I would prefer to only allow the user to withdraw the full amount inside the `withdrawal` record, not allow them to specify the amount.

Reply:
> Try can only be used with external function calls and contract creation calls.

Ah yes sorry, in this case just use `safeTransfer` which returns a boolean.

> Because this contract can never send ETH out except by calling the withdrawal function.

Just like ERC20 tokens no? Where is there a difference between ETH/ERC20? I thought this was for unexpected losses on the the deposit contract.

> I would prefer to only allow the user to withdraw the full amount inside the withdrawal record, not allow them to specify the amount.

Okay, makes sense.

Reply:
> Because this contract can never send ETH out except by calling the withdrawal function.

> Just like ERC20 tokens no? Where is there a difference between ETH/ERC20? I thought this was for unexpected losses on the deposit contract.

Or do you mean with ETH nothing unexpected can happen like on a token contract. That would indeed be true.

Reply:
please take another look